### PR TITLE
For sno-405

### PR DESCRIPTION
**Pull request by yrong**
Companion with https://github.com/Snowfork/cumulus/pull/8

### CODE REVIEW

**Comment by claravanstaden** [comment]
Hey @yrong! You are correct, in practice the weak subjectivity period is actually much longer than 27 hours. However, if I understand it correctly, that in the event where a massive number of validators get slashed, the subjectivity period could become shorter than the subjectivity period that we have set. Perhaps I also overemphasised the importance of [this statement](https://notes.ethereum.org/@adiasg/weak-subjectvity-eth2#:~:text=The%20maximum%20safe%20weak%20subjectivity%20period%20is%20always%20at%20least%2027%20hours).

I see that the article also takes into account a safety decay and the 2 weeks that you have set the weak subjectivity period is for about 200k validators, and on mainnet there are about 560k validators, so perhaps 2 weeks is a reasonable setting.

**Comment by claravanstaden** [comment]
With regards to the backfill, I suspect this will still be necessary, because if we don't do the backfill for the missing sync committee periods and beacon headers, we will not be able to verify messages that occurred in the offline period. But will defer to @vgeddes here if it needs to be done now. 😄

**Comment by yrong** [comment]
> With regards to the backfill, I suspect this will still be necessary, because if we don't do the backfill for the missing sync committee periods and beacon headers, we will not be able to verify messages that occurred in the offline period.

@claravanstaden Many thanks for the review! One misunderstanding I would like to address is that `backfill` what I refer to here is the reverse sync mechanism you described in notion page,  to make things easy at this moment what I would prefer is adding missing sync committee periods through unblock call directly as following.
https://github.com/Snowfork/snowbridge/blob/218cf483c365df8bbd05b1846881366bcb2ba57e/parachain/pallets/ethereum-beacon-client/src/lib.rs#L392-L410

Based on `Weak-Subjectivity` is extended to 14 days, then actually there is only a small chance that we need to do such backfill recovery.  For normal cases AFAIK our beacon relayer will just recover smoothly from breakpoint without intervention but correct me if I'm wrong.

Yes,  I'll wait to see more comments from team.

**Comment by claravanstaden** [comment]
@yrong I see! Sure, that makes sense. I guess we'd just have to get an easy way to provide all the checkpoints to governance through a script or something, but that is certainly more simple.

> Based on WeakSubjectivityPeriod is extended to 14 days, then actually there is only a small chance that we need to do such backfill recovery. For normal cases AFAIK our beacon relayer will just recover smoothly from breakpoint without intervention but correct me if I'm wrong.

Totally agree. 👌

File: parachain/pallets/ethereum-beacon-client/src/lib.rs
Comment (claravanstaden) [inline parachain/pallets/ethereum-beacon-client/src/lib.rs:976]: Nice! ✨

**Review by claravanstaden — COMMENTED** [review]
Awesome!

File: core/packages/test/scripts/fetch-beacon-data.sh
Comment (alistair-singh) [inline core/packages/test/scripts/fetch-beacon-data.sh:3]: Remove commented code.

File: parachain/pallets/ethereum-beacon-client/src/lib.rs
Comment (alistair-singh) [inline parachain/pallets/ethereum-beacon-client/src/lib.rs:378]: `process_checkpoint_sync` uses `verify_sync_committee` to verify checkpoint. My understanding is that it validates that the provided SyncCommittee exists within the beacon state at the header specified.

We provide a list of CheckPoints in `unblock_bridge` to process in order. It looks like we validate each checkpoint independently. We do not explicitly check that all these checkpoints lie on the canonical chain or any SyncCommitee handovers. Is this correct?

Will this be handled by governance in the sense that the community would make sure the checkpoints are correct before voting on a proposal that would enact the unblocking?

**Review by alistair-singh — COMMENTED** [review]
Looks good!

File: parachain/pallets/ethereum-beacon-client/src/lib.rs
Comment (yrong) [reply parachain/pallets/ethereum-beacon-client/src/lib.rs:380]: Exactly!

File: parachain/pallets/ethereum-beacon-client/src/lib.rs
Comment (yrong) [reply parachain/pallets/ethereum-beacon-client/src/lib.rs:372]: I would assume not nessessary all in in one extrinsic, multiple calls each with partial of the missing data should also be fine. 
And as mentioned above, with some monitoring components running(not scope of this PR) out-of-sync can be easily checked so that we can respond in time. For normal cases our beacon relayer will just recover smoothly from breakpoint without the `backfill`  here.

Comment (alistair-singh) [reply parachain/pallets/ethereum-beacon-client/src/lib.rs:372]: Will multiple calls mean multiple governance proposals?

Comment (yrong) [reply parachain/pallets/ethereum-beacon-client/src/lib.rs:372]: Probably. Not sure if one governance proposal which contains multiple calls(like `batch_all` but can be scheduled in multiple blocks) is possible or we'll need multiple proposals. And based on improvements introduced in [Gov2](https://medium.com/polkadot-network/gov2-polkadots-next-generation-of-decentralised-governance-4d9ef657d11b) multiple propsals/referendums can be fast-tracked and run simultaneously I would suspect it's not a big problem.

**Review by alistair-singh — APPROVED** [review]
🚀

File: parachain/pallets/ethereum-beacon-client/src/lib.rs
Comment (vgeddes) [reply parachain/pallets/ethereum-beacon-client/src/lib.rs:372]: I have an idea: The governance proposal would include the _hashes_ of the checkpoints.

The proposal would be the following extrinsic `begin_recovery`:

```rust
fn begin_recovery(checkpoints: Vec<CheckPointHash>);
```

Then any user can permissionlessly sync individual checkpoints, as long as they are in the list above:
```rust
fn sync_recovery_checkpoint(checkpoint: CheckPoint);
```

Once all checkpoints have been synced,  call this extrinsic to unblock the bridge:
```rust
fn complete_recovery();
```

Comment (yrong) [reply parachain/pallets/ethereum-beacon-client/src/lib.rs:372]: Cool, definitely more friendly for governance! I'll do some refactoring accordingly.

Comment (yrong) [reply parachain/pallets/ethereum-beacon-client/src/lib.rs]: @vgeddes https://github.com/Snowfork/snowbridge/pull/795/commits/3fbe0eb1af318e56b2fed9b87b0a531369bc04fa

File: parachain/pallets/ethereum-beacon-client/src/lib.rs
Comment (alistair-singh) [inline parachain/pallets/ethereum-beacon-client/src/lib.rs:419]: Discussion: We can unify the `process_initial_sync` and `unblock_bridge` code paths. Instead of having an initial sync step, we could start the bridge in a `Deactivated` state by default and then use the `unblock_bridge` extrinsic to bring it into an Activated state. This will simplify the deployment process as we will not have to set `initial_sync` field in the beacon client genesis state.

Comment (yrong) [reply parachain/pallets/ethereum-beacon-client/src/lib.rs:433]: Good point!  Some changes accordingly.
https://github.com/Snowfork/snowbridge/pull/795/commits/4b1279a84029f407945df500c767d56b6b52deb3

**Review by alistair-singh — APPROVED** [review]

File: core/packages/test/scripts/fetch-beacon-data.sh
Comment (claravanstaden) [inline core/packages/test/scripts/fetch-beacon-data.sh:31]: Why do we get the beacon state for local testnet only? 🤔

File: parachain/pallets/ethereum-beacon-client/src/lib.rs
Comment (claravanstaden) [reply parachain/pallets/ethereum-beacon-client/src/lib.rs:433]: I wonder if the methods shouldn't then mirror the language used, for example, rename `block_bridge` to `deactivate` and `unblock_bridge` to `activate`?

File: parachain/pallets/ethereum-beacon-client/src/lib.rs
Comment (claravanstaden) [inline parachain/pallets/ethereum-beacon-client/src/lib.rs:758]: We can remove this, I think I missed removing this while testing the ancestry proofs. 😄 

```suggestion
```

File: parachain/pallets/ethereum-beacon-client/src/lib.rs
Comment (claravanstaden) [inline parachain/pallets/ethereum-beacon-client/src/lib.rs:990]: Why does this method need to return a dispatch result? I don't see an error condition that can arise in the method, unless I am missing it?

File: parachain/primitives/beacon/src/lib.rs
Comment (claravanstaden) [inline parachain/primitives/beacon/src/lib.rs:190]: This is just for tests, right? Perhaps you can just comment it is only being used in the tests. 😸

File: core/packages/test/scripts/fetch-beacon-data.sh
Comment (yrong) [reply core/packages/test/scripts/fetch-beacon-data.sh:31]: Since we've provided [script](https://github.com/Snowfork/snowbridge/blob/ron/sno-405/core/packages/test/scripts/start-goerli.sh) to setup goerli nodes from locally I would suggest we fetch beacon state of goerli also in this way.
Btw, I’ve never successfully get beacon state from public [endpoint](https://lodestar-goerli.chainsafe.io) before, :joy:  guess related to the bad network in China.

File: parachain/primitives/beacon/src/lib.rs
Comment (yrong) [reply parachain/primitives/beacon/src/lib.rs:190]: No, actually we use it in `process_initial_sync` [here](https://github.com/Snowfork/snowbridge/blob/0176d783df795ddaf38491a8c22caca1523ca692/parachain/pallets/ethereum-beacon-client/src/lib.rs#L435-L437)  so to reuse the logic in `process_checkpoint_sync`, make code here a bit cleaner IMHO.

**Pull request closed without merging**
